ranking_models/full_additive_joint_ranking.py

```diff
@@ -0,0 +1,117 @@
+import torch.nn as nn
+import torch.nn.functional as F
+import utils.composition_functions as comp_functions
+import torch
+
+
+class FullAdditiveJointRanker(nn.Module):
+    """
+    This class contains the full additive composition model JOINT model. it trains two representations, each containing
+    a different type of semantic content (one for the general phrase and one to be close the the conveyed attribute).
+    Both representations are combined into one representation.
+    The model takes two words and returns their composed representation.
+    :param input_dim: embedding dimension
+    :param dropout_rate: dropout rate for regularization
+    :param normalize_embeddings: whether the composed representation should be normalized to unit length
+    """
+
+    def __init__(self, input_dim, dropout_rate, normalize_embeddings):
+        """
+        :param input_dim: embedding dimension
+        :param dropout_rate: dropout rate for regularization
+        :param normalize_embeddings: whether the composed representation should be normalized to unit length
+        """
+        super(FullAdditiveJointRanker, self).__init__()
+
+        self._general_adj_matrix = nn.Linear(input_dim, input_dim)
+        self._general_noun_matrix = nn.Linear(input_dim, input_dim)
+
```

Comment:
hmm jetzt haben wir es so:
1) linear transformation with bias of each positional vector (each vector gets transformed with a position-specific matrix)
2) task-specific linear transformation of each vector and combination of the two word vectors via sum
3) combination of task-specific representations via sum

mir ist aufgefallen:
- wir haben verschiedene intialisierungsmethoden (identity matrix für task-specific, random für general). identity matrix Initialisierungen machen vor allen Dingen sinn wenn ich die direkt auf die originalen Vektoren anwende, denn dann kann das Modell, wenn die originalen vectoren schon ohne transformation perfekt wären, nahe an der identity matrix bleiben. also wenn würde vielleicht erst identity und dann random sinn machen, wahrscheinlich aber wäre es gut wenn alle gleich initialisiert, also entweder random oder idenity würden
- da es im originalen identity ist, würde ich jetzt erstmal diese Initialisierung auswählen
- wir haben keine non-linearity (bei den anderen Modellen haben wir eine zwischen den transformationssteps), hier könnte man eine nach der general transformation einbauen
- du hattest jetzt sehr gute Ergebnisse ich habe keine Ahnung ob genau diese Optionen die aktuell da sind, zu diesen guten Ergebnissen führen und zb eine non-linearity nach der general transformation alles kaputt machen würde..also vielleicht kannst du auch einmal auf dem kleinsten Datenset vergleichen wir groß der unterschied mit / ohne relu / mit identity / random init ist...

Reply:
ansonsten sieht alles perfekt aus, man könnte es also mergen, wenn wir uns hier auf eine Einstellung einigen

Reply:
zu 1.) ja ich wusste nicht ob ich wirklich bias adden soll. Ich hab's einfach dann möglichst analog zum matrix joint model gemacht und da habe ich das eben auch gemacht.
Denkst du, ich soll das ändern? 

meinst du mit _original_ das full_additive model (ohne joint)? Ich weiß nicht was gescheiter ist. bei nn.Linear gibt's, soweit ich herausgefunden habe, keinen Initalisierungs-parameter. Ich könnte also das auch "manuell" machen (matrix multiplication) und falls wir bias behalten noch bias dazu addieren. 

wenn wir das full-additive joint möglichst nahe bei dem von Hartung haben wollen, wieso sollen wir dann eine non-linearity einfügen? Aber klar, ich kann das noch ausprobieren.

Reply:
dann lassen wir es linear. keine nonlinearity und kein bias. dann also 6 lineare Transformationen, alle mit identity initialisiert

Reply:
ich habe das jetzt gemacht.
Die Frage ist jetzt nur: soll ich das beim Matrix joint model auch noch ändern? und müsste es dann theoretisch beim matrix normalen (nicht joint) auch noch geändert werden?